lib/Conversion/StandardToHandshake/StandardToHandshake.cpp

```diff
@@ -1808,29 +1785,66 @@ FuncOpLowering::replaceCallOps(ConversionPatternRewriter &rewriter) {
   return success();
 }
 
-LogicalResult FuncOpLowering::finalize(ConversionPatternRewriter &rewriter,
-                                       TypeRange argTypes, TypeRange resTypes,
-                                       func::FuncOp origFunc) {
-  SmallVector<Type> newArgTypes(argTypes);
-  newArgTypes.push_back(rewriter.getNoneType());
+LogicalResult HandshakeLowering::finalize(ConversionPatternRewriter &rewriter) {
+  auto ctrlArg =
+      r.front().addArgument(rewriter.getNoneType(), rewriter.getUnknownLoc());
 
-  auto funcType = rewriter.getFunctionType(newArgTypes, resTypes);
-  f.setType(funcType);
+  Operation *startOp = findStartOp(r);
+  startOp->getResult(0).replaceAllUsesWith(ctrlArg);
+  rewriter.eraseOp(startOp);
 
-  if (!f.isExternal()) {
-    auto ctrlArg =
-        f.front().addArgument(rewriter.getNoneType(), rewriter.getUnknownLoc());
-
-    // We've now added all types to the handshake.funcOp; resolve arg- and
-    // res names to ensure they are up to date with the final type
-    // signature.
-    f.resolveArgAndResNames();
+  return success();
+}
 
-    Operation *startOp = findStartOp(&f.getRegion());
-    startOp->getResult(0).replaceAllUsesWith(ctrlArg);
-    rewriter.eraseOp(startOp);
+LogicalResult handshake::lowerRegion(HandshakeLowering &hl,
```

Comment:
I was just asking myself if we could skip all the additional pattern matches and just pass in the rewriter that the original match on the parent operation produces. 
In this case here, one might be able to write one pattern for `FuncOp`, do the initial processing and then pass in the rewriter to this helper function.

Reply:
I played around with that locally and it seems to work for almost everything, except for the `checkDataflowConversion` function. 
Other functions, e.g, `connectConstantsToControl`, replace operations but these changes aren't directly visible to later functions during the same partial lowering. Therefore, the Values produced by `handshake::ConstantOp`s have zero uses when the `checkDataflowConversion` function is executed. 

I would argue that this check might be in the wrong place, as dialect invariant might be invalidated during a rewrite but should be re-established when a conversion finishes, which it does.

Reply:
nvm, all rewrites that change values are also broken. This is related to the limitation I mentioned.